From 061f8871a6b4d88943c2b50438ddd57c7c7e2b0f Mon Sep 17 00:00:00 2001
From: Sammy He <r62914@freescale.com>
Date: Wed, 4 Nov 2009 17:27:26 +0800
Subject: [PATCH 127/505] ENGR00118035 VPU: Add vpu software reset processing
 in unit test

Add vpu_SWReset() function call if one frame cannot finish decoding
in 5 seconds.

Signed-off-by: Sammy He <r62914@freescale.com>
---
 test/mxc_vpu_test/dec.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/test/mxc_vpu_test/dec.c b/test/mxc_vpu_test/dec.c
index 9f75561..7d0c777 100644
--- a/test/mxc_vpu_test/dec.c
+++ b/test/mxc_vpu_test/dec.c
@@ -659,7 +659,7 @@ decoder_start(struct decode *dec)
 	int err, eos = 0, fill_end_bs = 0, decodefinish = 0;
 	struct timeval tdec_begin,tdec_end, total_start, total_end;
 	RetCode ret;
-	int sec, usec;
+	int sec, usec, loop_id;
 	u32 yuv_addr, img_size;
 	double tdec_time = 0, frame_id = 0, total_time=0;
 	int decIndex = 0;
@@ -814,6 +814,7 @@ decoder_start(struct decode *dec)
 		}
 
 		is_waited_int = 0;
+		loop_id = 0;
 		while (vpu_IsBusy()) {
 			err = dec_fill_bsbuffer(handle, dec->cmdl,
 				      dec->virt_bsbuf_addr,
@@ -826,9 +827,22 @@ decoder_start(struct decode *dec)
 				return -1;
 			}
 
+			/*
+			 * Suppose vpu is hang if one frame cannot be decoded in 5s,
+			 * then do vpu software reset.
+			 * Please DON'T add the code if the application is in network
+			 * case since vpu interrupt also cannot be received if no
+			 * enough data.
+			 */
+			if (loop_id == 10) {
+				err = vpu_SWReset(handle, 0);
+				return -1;
+				}
+
 			if (!err) {
 				vpu_WaitForInt(500);
 				is_waited_int = 1;
+				loop_id ++;
 			}
 		}
 
-- 
1.8.0

